files/galaxy-test/dynamic_rules/usegalaxy/sorting_hat.py

```diff
@@ -308,40 +365,59 @@ def convert_to(tool_spec, runner):
     return tool_spec
 
 
-def _gateway(tool_id, user_roles, user_id, user_email, memory_scale=1.0):
+def _gateway(tool_id, user_preferences, user_roles, user_id, user_email, memory_scale=1.0):
     tool_spec = _finalize_tool_spec(tool_id, user_roles, memory_scale=memory_scale)
 
     # Now build the full spec
     runner_hint = None
 
-    if tool_id != 'upload1':
-        hints = [x for x in user_roles if x.startswith('destination-')]
-        if len(hints) > 0:
-            runner_hint = hints[0].replace('destination-pulsar-', 'remote_cluster_mq_')
+    if tool_id not in ('upload1', '__DATA_FETCH__', '__SET_METADATA__'):
+        # hints = [x for x in user_roles if x.startswith('destination-')]
+        # if len(hints) > 0:
+        #     runner_hint = hints[0].replace('destination-pulsar-', 'remote_cluster_mq_')
+        for data_item in user_preferences:
+            if "distributed_compute|remote_resources" in data_item:
+                if user_preferences[data_item] != "None":
+                    runner_hint = user_preferences[data_item]
 
     # Ensure that this tool is permitted to run, otherwise, throw an exception.
     assert_permissions(tool_spec, user_email, user_roles)
 
-    env, params, runner, _ = build_spec(tool_spec, runner_hint=runner_hint)
+    env, params, runner, _, tags = build_spec(tool_spec, runner_hint=runner_hint)
     params['accounting_group_user'] = str(user_id)
     params['description'] = get_tool_id(tool_id)
 
-    return env, params, runner, tool_spec
+    # This is a special case, we're requiring it for faster feedback / turnaround times.
+    if 'training-hard-limits' in user_roles:
+        params['requirements'] = 'GalaxyGroup  ==  "training-hard-limits"'
+
+    # Shortcut to run jobs only on Py3 enables nodes
+    if 'py3' in user_roles:
```

Comment:
Does this mean we need to rebuild VGCN with py3k?

Reply:
No

Reply:
I have prepared two workers that are mounting the test venv dir; the code above is just a shortcut to be sure jobs from test run on these nodes only.